openapi.yaml

```diff
@@ -4711,16 +4711,31 @@ components:
       example: null
       description: |-
         The namespace the `process_id` is valid for.
+
+        The following options are pre-defined by the openEO API, but additional
+        namespaces may be introduced by back-ends or in a future version of the API.
         
-        By default, the namespace is `null`, which includes all pre-defined processes listed at `GET /processes`.
-        The namespace `user` refers to all user-defined processes listed as `GET /process_graphs`.
-        Additional namespaces may be introduced by back-ends or in a future version of the API.
+        * `null` (default): Checks both user-defined and pre-defined processes,
+           but prefers user-defined processes if both are available.
+           This allows users to replace missing pre-defined processes with
+           user-defined processes for portability, e.g. common processes from
+           [processes.openeo.org](https://processes.openeo.org) that have a process graph
+           included.
+           If a user-defined process is prioritized over a pre-defined process,
+           a log entry SHOULD be added.
+           To ensure reproducibility, back-ends SHOULD persist the namespace that
+           is prioritized at the time the process is submitted so that adding or
+           removing processes doesn't change the processing results in the future.
```

Comment:
> To ensure reproducibility, backend SHOULD persist ...so that .. doesn't chang the processing results ...

This  reads a bit vague to me, but it still sounds that there is quite some responsibility for a backend regarding reproducibility features. Like an old process graph should still run in some kind of snapshotted UDP/PDP context?

Reply:
also if you speak about persisting: persist based on what handle or identifier? for example: jobs and services have an  identifier, but for synchronous `/result` call there is no identifier .

Reply:
Yes, this is more a recommendation to the back-end though. At least I'd expect that at least PDP give the same results after a year and also UDP unless explicitly altered by a user. If a PDP changes, you are lost as a user anyway, with or without namespaces.

Reply:
Maybe it should read:

```> To *improve* reproducibility, it is RECOMMENDED that back-ends persist the namespace that
           is prioritized (selected?) at the time the process is submitted so that adding or
           removing processes doesn't change the processing results in the future.

Reply:
Made some changes in the last commit to not put too much pressure regarding reproducibility.

Reply:
Wouldn't it just be easier to say that if reproducibility is important for the user it is recommended to specify the namespace explicitly instead of depending on run-time name resolution details?

Reply:
User usually don't read the API spec and as a user I'd expect something like that from the back-end tbh.

Reply:
Also, another sidenote: I don't think that UDP are the best tool for users that are very wary about reproducibility: you only have a single user-level global UDP namespace. In modern development it's typically recommended to work in per-project sandboxes (e.g. virtual envs in python, etc). 
I'm not saying we should allow users to define multiple namespaces (yet :smile:  ), but if I would need a highly reproducible and stable workflow with openEO, I think I would prefer to just embed everything in a large PG instead of managing and calling UDPs.

Reply:
Personally, I'm not a huge fan of this time-based namespace resolution and am even less of a fan that it is a recommendation only. In practice, a user then can't be sure what's happening or if the same approach is used on two different back-ends.
As a user I'd prefer to having to add/update a namespace in my PG and hence ensure that the back-end will run the job I want it to run, rather than leaving the namespace blank and crossing my fingers..

Reply:
How would you update that? Except for the Web Editor there's no client AFAIK that allows to generate code from a PG, so a user would need to edit the PG, which is not something we want to reply on.

Reply:
> I'm not saying we should allow users to define multiple namespaces (yet 😄 ), but if I would need a highly reproducible and stable workflow with openEO, I think I would prefer to just embed everything in a large PG instead of managing and calling UDPs.

Indeed, it would probably be a good idea if clients could just "resolve" UDP and create on big PG to be submitted. Then you still rely on PDP to not change, but maybe versioned PDP become a thing at some point 😄

Reply:
with the Python client (and I guess the R client too) you express the PG in Python code, so it's trivial to edit/update that.

Also at VITO we have a user that doesn't want to depend too much on the Python client because it's not stable enough: he just does HTTP calls directly (from python), and manages a JSON file somewhere

Reply:
Also: if you're using openEO in a serious way, at some point you probably are going to want to download some kind of representation of the PG in some way (JSON probably), for example to put it in version control or attach it to a report. So I don't think we should shield users too much from JSON. (If it would be XML, that would be another story :smile: )

Reply:
> with the Python client (and I guess the R client too) you express the PG in Python code, so it's trivial to edit/update that.

Short term, yes, usually. Long term, probably less likely. And very unlikely for shared processes, which is the main reason to allow the `null` namespace.

> Also at VITO we have a user that doesn't want to depend too much on the Python client because it's not stable enough: he just does HTTP calls directly (from python), and manages a JSON file somewhere

Interesting, but I doubt that's not the normal use case ;-)

Reply:
imho the clients must set the namespace explicitly. 
When the user connects to a back-end, the clients query /processes and /process_graphs to get the PDPs and UDPs available at the back-end for that user. If the user selects a process for which there is a clash, the user should be asked to flag which process to use. The clients then sets the namespace accordingly before submitting the PG to the back-end.

Reply:
I'm not against setting it explicitly when there's a clash, but unless there is a clash, the namespace should not be set.
The namespace should not be set by default so that sharing and importing PGs works better.

Reply:
> The clients then sets the namespace accordingly before  ...

that would be the most user friendly, but you still need to specify back-end side behavior because there might be users that don't use (official) clients.

Reply:
So should I add:

> Clients SHOULD ask the user to selects the namespace, if multiple processes with the same name exist.

Reply:
> And very unlikely for shared processes (e.g. Hub),

To me that doesn't sound like a big problem. When I copy-paste code snippets from, say stackoverflow or some blog, I don't expect it to work automagically. There is always some  adaptation to do (changing variable names, fixing imports, ...).  We have this problem anyway with collection ids for example.

Reply:
But that's what we made UDP for, that they can easily be shared. Like a library. You usually don't change libraries. I'm not really thinking about code snippets here, but libraries. And Collection IDs in UDPs should be parameters.

Reply:
> Clients SHOULD ask the user to selects

that sound quite oriented towards interactive clients like the web editor. That doesn't work nice for clients that are expected to work in some kind of headless/batch mode (e.g. python scripts)

Reply:
> that sound quite oriented towards interactive clients like the web editor. That doesn't work nice for clients that are expected to work in some kind of headless/batch mode (e.g. python scripts)

Why is that a problem?

> If you have ndvi as UDP and PDP and you call let's say builder$ndvi(data) then it prints/throws (whatever you like) a warning and could recommends you to set the namespace to null, user or backend. Should work in R, I guess also at least in Python Notebooks?

Reply:
UDPs are a bit like libraries, but a lot more individual and granualar than how libraries work in other platforms.
I think a better equivalent of "libararies" in openeEO would be a using a URL as namespace, pointing to a repository of reusable processes.

(which is an approach we are thinking about at VITO for a use case we started working on)

Reply:
Okay, maybe it's more like a library of functions, but you also don't go ahead and change functions in a library. You just use them according to the contract/interface...

Reply:
"user-defined" does sounds like I have permissions to do what I want

Reply:
Sure, you can submit whatever you like. You can also fork libraries on GitHub, but that's not how they are usually used. 

If you go to the Hub and there's a process you'd like to use, it's just a process that you can submit as a "user-defined" process to your back-end.

Reply:
thinking a bit out of the box here: could we add some kind of "name resolution strategy" property somewhere (e.g. as parameter in job/service/result request, or as field in `process` JSON body), so that user/client can pick a strategy that is best for the usecase or context?

Reply:
> yes, you can use python client in interactive contexts. But my point is that if I have a script which I run daily with some cron system, I don't want it to suddenly crash/hang one day because the backend introduced a "collision"

Okay, maybe it's better to say:
> If multiple processes with the same identifier exist, Clients SHOULD inform the user that it's recommended to select a namespace.

Then it shouldn't crash

Reply:
> thinking a bit out of the box here: could we add some kind of "name resolution strategy" property somewhere (e.g. as parameter in job/service/result request, or as field in `process` JSON body), so that user/client can pick a strategy that is best for the usecase or context?

So that it would allow to let's say prefer PDP over UDP?

Reply:
indeed, 
and in the longer term if more namespaces are defined (officially or customly by a backend): in which order to scan the namespaces

Reply:
for example: it's a list like `["user", "backend"]` to say to prefer UDP over backend. and if the list is empty or null: you error on any collision

Reply:
I thought you were aiming to simplify and remove the default behavior, but that sounds even more complex and time-based ;-)  There are only really UDP and PDP in the resolution strategy, all others are likely having namespaces (usually URLs) anyway, I guess.

Reply:
and it would only apply for processes where namespace is not set

Reply:
Okay, hmm... I don't really see the advantage, I must be missing something.
There's only really three cases: prefer UDP, prefer PDP and error
Currently we do prefer UDP and I thought you were not liking error. Prefer PDP currently is only possible by setting the namespace for a new process or removing the UDP. Does that really need a resolution strategy parameter?

Reply:
there is still the edge case where a user has multiple UDPs in his/her PG, some of which now clash with PDPs, and they want only a subset of them to be used as UDP. In this case they must specify the namespace explicitly.

Reply:
> all others are likely having namespaces

Indeed that's true currently. However, it could be extended to for example `"name_resolution_order"=["http://url/to/repo", "backend"]`, which would mean that processes without an explicit namespace are first looked up at that url.

> but that sounds even more complex and time-based

true, it indeed smells like feature creep. Just trying to find a way out of this ~~conflict~~ issue :smile:

Reply:
I guess we can't solve all edge cases, but at least they can specify the namespace per process then.

Reply:
> Currently we do prefer UDP and I thought you were not liking error by default

I don't like automagic name resolution by default. And I don't like the introduction of "conflict" as a thing to worry about. 

The current behavior per API without PR 313 is: use "backend" namespace by default if namespace is not set, and if it is set: use that. There is not concept of conflict whatsoever. And when the process can not be found: you have an error. Plain and simple

Reply:
This PR 313 introduces the concepts "name resolution" and "name conflict" for reasons and use cases I'm not that famliar with, but for which my simple solution would be "just specify the namespace explicitly".

Reply:
some reasons I don't like (automatic) name resolution:
- it is dependent on timing, user specific state (their set of UDPs), which makes thing harder to investigate and reason about
- we don't care about overloading rules or preserving the signature (input and output types) of a function, which should be part I think of a dynamic resolution system

Reply:
> issue/conflict: how does the back-end resolve which process to use when `namespace` is null and there are one UDP and one PDP with the same id?

Currently it says UDP over PDP as it is most flexible: Users have two of options to avoid that: delete UDP or set namespace.

> Possible solution from @soxofaan : specify resolution_order as a parameter. If missing, return conflict errors

What is the advantage and what should be the default?

> This approach is not bad because it makes it explicit for back-ends which process to use (as oppposed to finding out which was set first between UDP and PDP)

What you describe here is not the behavior proposed in this PR. The PR doesn't say you need to find out what was first, it just says prefer UDP over PDP. That seems exactly as explicit?!

> I don't like automagic name resolution by default. And I don't like the introduction of "conflict" as a thing to worry about.

> The current behavior per API without PR 313 is: use "backend" namespace by default if namespace is not set, and if it is set: use that. There is not concept of conflict whatsoever. And when the process can not be found: you have an error. Plain and simple

Yes, but we forgot a use-case in the PR unfortunately. We introduced the process graphs to openeo-processes so that they can be used instead of missing pre-defined processes at back-ends. And I also didn't really took sharing into account there. So I'm now just trying to get (partly) back to RC2. There were means to avoid collisions in RC2 (reject UDP, prefer UDP oder (later introduced) PDP). But I removed that as you requested it, but now you don't want the conflicts, too. That's a bit confusing ;-)